src/hypothesis/strategies.py

```diff
@@ -412,13 +413,17 @@ def sampled_from(elements):
 
     from hypothesis.searchstrategy.misc import SampledFromStrategy, \
         JustStrategy
-    elements = tuple(iter(elements))
+    from hypothesis.internal.conjecture.utils import check_sample
     if not elements:
         return nothing()
     if len(elements) == 1:
-        return JustStrategy(elements[0])
-    else:
-        return SampledFromStrategy(elements)
+        return JustStrategy(next(iter(elements)))
+    try:
+        elements = check_sample(elements)
+    except InvalidArgument as err:
+        elements = tuple(elements)
+        note_deprecation(err.args[0] if err.args else u'Unorderable elements')
```

Comment:
Would it make sense to reraise if we're currently in strict mode (where deprecations are errors anyway)? It's a bit different from what we do elsewhere, but it feels weird to just reraise with a deprecation warning in that case.